File: Grind75/Week One/valid_parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */

// First Solution:
var isValid = function(s) {
    
    // s.length() == 0 ||
    if ( s.length % 2 != 0 ){
        return false;
    }

    let bracesStack = [];

    for (let str of s){
        

        if("{[(".includes(str) ){
            bracesStack.push(str);
        }

        if("}])".includes(str)){
            if (str == "}"){
                if(bracesStack.pop() != "{") {
                    return false;
                } 
            }
            if (str == "]"){
                if(bracesStack.pop() != "[") {
                    return false;
                } 
            }
            if (str == ")"){
                if(bracesStack.pop() != "(") {
                    return false;
                } 
            }
        }
    }
    if(bracesStack.length > 0){
        return false;
    }
    return true;

};

File: Grind75/Week One/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if (len(s) % 2 != 0):
            return False

        braces_stack = []

        for symbol in s:
            if symbol in "{[(":
                braces_stack.append(symbol)

                    
            if(symbol in ")]}" ):
                if(len(braces_stack) == 0):
                    return False
                    
                if(symbol is "}"):
                    if(braces_stack.pop() != "{"):
                        return False

                if(symbol is "]"):
                    if(braces_stack.pop() != "["):
                        return False

                if(symbol is ")"):
                    if(braces_stack.pop() != "("):
                        return False


        if(len(braces_stack) > 0):
            return False
        
        return True
    

